internal/sync2/pipe.go

```diff
@@ -0,0 +1,280 @@
+}
+
+// CloseWithError implements closing with error
+func (writer pipeWriter) CloseWithError(err error) error {
+	pipe := writer.pipe
+	pipe.mu.Lock()
+	if pipe.writerDone {
+		pipe.mu.Unlock()
+		return io.ErrClosedPipe
+	}
+	pipe.writerDone = true
+	pipe.writerErr = err
+	pipe.nodata.Broadcast()
+	pipe.mu.Unlock()
+
+	return pipe.closeHalf()
+}
+
+// closeHalf closes one side of the pipe
+func (pipe *pipe) closeHalf() error {
+	if atomic.AddInt32(&pipe.refcount, -1) == 0 {
+		return pipe.buffer.Close()
+	}
+	return nil
+}
+
+// Write writes to the pipe returning io.ErrClosedPipe when blockSize is reached
+func (writer pipeWriter) Write(data []byte) (n int, err error) {
+	pipe := writer.pipe
+	pipe.mu.Lock()
+
+	// has the reader finished?
+	if pipe.readerDone {
+		pipe.mu.Unlock()
+		return 0, pipe.readerErr
+	}
+
+	// have we closed already
+	if pipe.writerDone {
+		pipe.mu.Unlock()
+		return 0, io.ErrClosedPipe
+	}
+
+	// check how much do they want to write
+	canWrite := pipe.limit - pipe.write
+
+	// no more room to write
+	if canWrite == 0 {
+		pipe.mu.Unlock()
+		return 0, io.ErrClosedPipe
+	}
+
+	// figure out how much to write
+	toWrite := int64(len(data))
+	if toWrite > canWrite {
+		toWrite = canWrite
+	}
+
+	writeAt := pipe.write
+	pipe.mu.Unlock()
+
+	// write data to buffer
+	writeAmount, err := pipe.buffer.WriteAt(data[:toWrite], writeAt)
+
+	pipe.mu.Lock()
+	// update writing head
+	pipe.write += int64(writeAmount)
+	// wake up reader
+	pipe.nodata.Broadcast()
+	// check whether we have finished
+	done := pipe.write >= pipe.limit
+	pipe.mu.Unlock()
+
+	if err == nil && done {
+		err = io.ErrClosedPipe
+	}
+	return writeAmount, err
+}
+
+// Read reads from the pipe returning io.EOF when writer is closed or blockSize is reached
+func (reader pipeReader) Read(data []byte) (n int, err error) {
+	pipe := reader.pipe
+	pipe.mu.Lock()
+	// wait until we have something to read
+	for pipe.read >= pipe.write {
+		// has the writer finished?
+		if pipe.writerDone {
+			pipe.mu.Unlock()
+			return 0, pipe.writerErr
+		}
+
+		// have we closed already
+		if pipe.readerDone {
+			pipe.mu.Unlock()
+			return 0, io.ErrClosedPipe
+		}
+
+		// have we run out of the limit
+		if pipe.read >= pipe.limit {
+			pipe.mu.Unlock()
+			return 0, io.EOF
+		}
+
+		// ok, lets wait
+		pipe.nodata.Wait()
+	}
+
+	// how much there's available for reading
+	canRead := pipe.write - pipe.read
+	// how much do they want to read?
+	toRead := int64(len(data))
+	if toRead > canRead {
+		toRead = canRead
+	}
+	readAt := pipe.read
+	pipe.mu.Unlock()
+
+	// read data
+	readAmount, err := pipe.buffer.ReadAt(data[:toRead], readAt)
+
+	pipe.mu.Lock()
+	// update info on how much we have read
+	pipe.read += int64(readAmount)
+	done := pipe.read >= pipe.limit
+	pipe.mu.Unlock()
+
+	if err == nil && done {
+		err = io.EOF
+	}
+	return readAmount, err
+}
+
+// MultiPipe is a multipipe backed by a single file
+type MultiPipe struct {
+	pipes    []pipe
+	refcount int64
```

Comment:
What's the purpose of this field? I'm not sure if I'm missing something.

Reply:
In general I know what its doing but I'm not sure about its meaning in `MultiPipie`. Its just set to `blockCount` value and used to create `offsetFile`.

Reply:
The meaning in `MultiPipe` is the number of pipes that are still open (starts with the number of pipes specified). The meaning in `pipe` is the number of half-pipes that are open (starts with 2).

Maybe renaming the `blockCount` to `pipeCount` and `blockSize` to `pipeSize` would make it clearer?

Reply:
All pipes in the multi pipe share the same underlying resource (i.e. file). This is why the refcount is also shared between all the pipes.

Reply:
Also maybe rename `refcount` to `open` or `openPipes`.

Reply:
Ok, thanks for clarifications. Yes, using `pipies` in names should make it clearer.